feat(benchmarks) Add LLM evaluation pipeline for Medical challenge (#3768)
adap/flower: benchmarks/flowertune-llm/evaluation/medical/README.md

# Evaluation for Medical challenge

We build up a medical question answering (QA) pipeline to evaluate our fined-tuned LLMs.
Three datasets have been selected for this evaluation: [PubMedQA](https://huggingface.co/datasets/bigbio/pubmed_qa), [MedMCQA](https://huggingface.co/datasets/medmcqa), and [MedQA](https://huggingface.co/datasets/bigbio/med_qa). 

## Environment Setup

```shell
git clone --depth=1 https://github.com/adap/flower.git && mv flower/benchmarks/flowertune-llm/evaluation/medical ./flowertune-eval-medical && rm -rf flower && cd flowertune-eval-medical
```

Create a new Python environment (we recommend Python 3.10), activate it, then install dependencies with:

```shell
# From a new python environment, run:
pip install -r requirements.txt

# Log in HuggingFace account
huggingface-cli login
```

## Generate model decision & calculate accuracy

```bash

> python eval.py \
> --peft-path=/path/to/fine-tuned-peft-model-dir/ # e.g., ./peft_1
> --run-name=fl  # specified name for this run  
> --batch-size=16 
> --quantization=4 
> --datasets=pubmedqa,medmcqa,medqa

Reviewer: ```suggestion
python eval.py \
--peft-path=/path/to/fine-tuned-peft-model-dir/ \ # e.g., ./peft_1
--run-name=fl  \ # specified name for this run  
--batch-size=16 \
--quantization=4 \
--datasets=pubmedqa,medmcqa,medqa
```

## Evaluation for Medical challenge

> We leverage the medical question answering (QA) metric provided by [Meditron](https://github.com/epfLLM/meditron/tree/main/evaluation) to evaluate our fined-tuned LLMs.

Reviewer: ```suggestion
We leverage the medical question answering (QA) metric provided by [Meditron](https://github.com/epfLLM/meditron) to evaluate our fined-tuned LLMs.
```

Reviewer: I think it's better to link to the main readme

Three datasets have been selected for this evaluation: [PubMedQA](https://huggingface.co/datasets/bigbio/pubmed_qa), [MedMCQA](https://huggingface.co/datasets/medmcqa), and [MedQA](https://huggingface.co/datasets/bigbio/med_qa). 

### Step 0. Set up Environment

```shell
git clone --depth=1 https://github.com/adap/flower.git && mv flower/benchmarks/flowertune-llm/evaluation/medical ./flowertune-eval-medical && rm -rf flower && cd flowertune-eval-medical
```

Then, install dependencies with:

```shell
# From a new python environment, run:
pip install -e .

# Log in HuggingFace account
huggingface-cli login
```

### Step 1. Generate model answers to medical questions

> 

Reviewer: > [!NOTE]
> Evaluation needs to be run on PubMedQA, MedMCQA and MedQA

```bash
python inference.py \
--peft-path=/path/to/fine-tuned-peft-model-dir/  # e.g., ./peft_1
--dataset-name=pubmedqa  # chosen from [pubmedqa, medmcqa, medqa]

> --run-name=fl  # arbitrary name for this run 

Reviewer: ```suggestion
--run-name="fl-pubmedqa"  # an identifier for this run (up to you to choose) 
```

```
The answers will be saved to `benchmarks/generations/[dataset_name]-[run_name].jsonl` in default.

> ### Step 2. Calculate accuracy

Reviewer: ```suggestion
## Calculate accuracy
```

```bash
python evaluate.py \
--dataset-name=pubmedqa  # chosen from [pubmedqa, medmcqa, medqa]

> --run-name=fl  # run_name used in Step 1

Reviewer: ```suggestion
--run-name="fl-pubmedqa"  # run_name used in Step 1
```

```
The accuracy value will be printed on the screen.

> [!NOTE]

> > Please ensure that you provide all **three accuracy values** for three evaluation datasets when submitting to the LLM Leaderboard (see the [`Make Submission`](https://github.com/adap/flower/tree/main/benchmarks/flowertune-llm/evaluation#make-submission-on-flowertune-llm-leaderboard) section).

Reviewer: ```suggestion
> Please ensure that you provide all **three accuracy values (PubMedQA, MedMCQA, MedQA)** for three evaluation datasets when submitting to the LLM Leaderboard (see the [`Make Submission`](https://github.com/adap/flower/tree/main/benchmarks/flowertune-llm/evaluation#make-submission-on-flowertune-llm-leaderboard) section).
```
